Add fsync support for Distributed engine.

azat opened the pull request:
Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Two new settings (by analogy with MergeTree family) has been added:
- `fsync_after_insert` - Do fsync for every inserted. Will decreases performance of inserts.
- `fsync_directories` - Do fsync for temporary directory (that is used for async INSERT only) after all operations (writes, renames, etc.).

Detailed description / Documentation draft:
Now Distributed engine support the SETTINGS clause (since there are too much arguments already, and some of them will be converted to the SETTINGS Later), and two new settings, that listed above, added.

Refs: #17380 (p1)

azat committed:
Add fsync support for Distributed engine.

Two new settings (by analogy with MergeTree family) has been added:

- `fsync_after_insert` - Do fsync for every inserted. Will decreases
  performance of inserts.

- `fsync_tmp_directory` - Do fsync for temporary directory (that is used
  for async INSERT only) after all part operations (writes, renames,
  etc.).

Refs: #17380 (p1)

filimonov commented on src/Storages/Distributed/DistributedSettings.h:
```diff
@@ -0,0 +1,31 @@
+
+#define LIST_OF_DISTRIBUTED_SETTINGS(M) \
+    M(Bool, fsync_after_insert, false, "Do fsync for every inserted. Will decreases performance of inserts (only for async INSERT, i.e. insert_distributed_sync=false)", 0) \
+    M(Bool, fsync_tmp_directory, false, "Do fsync for temporary directory (that is used for async INSERT only) after all part operations (writes, renames, etc.).", 0) \
```
I have mixed feeling here... From one side i really like the idea of having Distributed-table related settings stored with table, from the other - we have already dozens of Distributed table settings stored as user-level settings

Maybe we can start moving there here, and some common pattern like "user settings will be used unless table settings redifine it". 

Maybe it's worth to add a test with ALTER TABLE ... MODIFY SETTINGS ...?

azat replied:
>we have already dozens of Distributed table settings stored as user-level settings

Some of them will be migrated to SETTINGS (there is TODO for this)

>Maybe we can start moving them here, and use some common pattern like "user settings will be used unless table settings redifine it"?

Yes, but separately, since it is not related to fsync.

>Maybe it's worth to add a test with ALTER TABLE ... MODIFY SETTINGS ...?

Right now `ALTER MODIFY SETTING` is not support

filimonov replied:
:+1:

filimonov commented on src/Storages/Distributed/DistributedBlockOutputStream.cpp:
```diff
@@ -622,6 +628,13 @@ void DistributedBlockOutputStream::writeToShard(const Block & block, const std::
 
         /// Write batch to temporary location
         {
+            std::optional<DirectorySyncGuard> tmp_path_sync_guard;
```
will it be recreated on every iteration?

azat replied:
There is no loop, there is just a sub scope (for objects lifetime)

filimonov commented on docs/en/engines/table-engines/special/distributed.md:
```diff
@@ -25,10 +25,19 @@ The Distributed engine accepts parameters:
+
+- `fsync_after_insert` - Do fsync for every inserted. Will decreases performance of inserts (only for async INSERT, i.e. `insert_distributed_sync=false`),
+
+- `fsync_tmp_directory` - Do fsync for temporary directory (that is used for async INSERT only) after all part operations (writes, renames, etc.).
```
Difference of those 2 setting is not very clear. Better description highlighting the difference is needed.

azat replied:
An suggestion?

filimonov replied:
I'm wondering if there is any use case when one fsync should be enables while the other - not. 

I guess the people who want guarantees and ok to make inserts slower, don't expect to get half-guarantee.

azat replied:
>I'm wondering if there is any use case when one fsync should be enabled while the other - not.

Most likely people tend to enable fsync but not fsync for directories
Also it varies from the underlying filesystem.

filimonov replied:
> Most likely people tend to enable fsync but not fsync for directories

Questionable, but OK. 

> An suggestion?

See the comment below.

filimonov commented on src/Storages/Distributed/DistributedBlockOutputStream.cpp:
```diff
@@ -647,6 +660,10 @@ void DistributedBlockOutputStream::writeToShard(const Block & block, const std::
+                out.sync();
         }
 
         // Create hardlink here to reuse increment number
```
shouldn't fsync on `path` (i.e. all dir_names) also happen to ensure that hardlink appeared in the folder?

azat replied:
This will not make any additional guarantee since background thread can already pick it for sending, the only file that can be protected so is file in the `tmp/` directory and the `tmp/` directory itself.

So there is still risk of data loss.

filimonov replied:
similar question about removing the file by the StorageDistributedDirectoryMonitor

filimonov replied:
> This will not make any additional guarantee since background thread can already pick it for sending, 

Even if distributedmonitor will be fast enough to consume the whole file (and even remove that already) it seems nothing bad will happen if we just do fsync on the folder here.

azat replied:
Nothing bad, it may mislead the reader of the code that this place is safe, but let's add it with a comment that it is not.

filimonov replied:
Yes, it will not be atomic and can leave the dirty state. But it will never return success to the user if something was not finished successfully

azat replied:
Yes, I was talking about that INSERT failures does not means that nothing was done, but this is harder to fix
I will get back fsync for other directories

filimonov replied:
👍

filimonov commented on docs/en/engines/table-engines/special/distributed.md:
```diff
@@ -25,10 +25,19 @@ The Distributed engine accepts parameters:
     -   [MergeTree](../../../engines/table-engines/mergetree-family/mergetree.md#table_engine-mergetree-multiple-volumes) for the examples
 
+Also it accept the following settings:
+
+- `fsync_after_insert` - Do fsync for every inserted. Will decreases performance of inserts (only for async INSERT, i.e. `insert_distributed_sync=false`),
+
+- `fsync_tmp_directory` - Do fsync for temporary directory (that is used for async INSERT only) after all part operations (writes, renames, etc.).
```
```suggestion
- `fsync_after_insert` - do the `fsync` for the file data after async insert. Guarantees that the OS flushed the whole file to a disk **on the initiator node**. 
- `fsync_tmp_directory` - do the `fsync` for directories after async insert. Guarantees that the OS refreshed directory metadata **on the initiator node** disk.

!!! note "Note"
    Durability settings (`fsync_...`):
    - Affect only async INSERTs (i.e. `insert_distributed_sync=false`) when data first stored on the initiator node disk and later asynchronously send to shards.
    - May significantly decrease the inserts' performance
    - Affect writing the data stored inside Distributed table folder into the **node which accepted your insert**. If you need to have guarantees of writing data to underlying MergeTree tables - see durability settings (`...fsync...`) in `system.merge_tree_settings`.
```

filimonov replied:
May be we should call the setting just `fsync_directories` ?

azat replied:
Thanks for the suggestion, LGTM, although I did not want to include "OS" word into the description, but I guess it is fine (anyway not it looks way more understandable for users)

azat replied:
>May be we should call the setting just fsync_directories or fsync_after_directory_operations ?

It was named by analogy to the MergeTree table, i.e. `fsync_part_directory`, I'm ok with both (actually even fsync_after_insert looks too verbose to me)

filimonov replied:
> did not want to include "OS" word into the description

Initially i also was not using that, but grammarly recommended me to rephrase to remove passive voice ('guarantee part were flushed')

> I'm ok with both 

I like `fsync_directories`.

filimonov committed:
Update documentation for Distributed fsync settings.

azat committed:
Sync other temporary directories for Distributed fsync_tmp_directories

azat committed:
Add fsync support for Distributed/DirectoryMonitor

Note that there is no fsync_tmp_directory support in DirectoryMonitor
since you cannot propagate the error to user anyway.

azat committed:
Accept IDisk in DirectoryMonitor (for further fsync)

azat committed:
Add fsync_tmp_directory support into DirectoryMonitor

azat committed:
Simple coverage of fsync_after_insert/fsync_tmp_directory with batching

filimonov reviewed (approved):
The name of the setting should be adjusted, the rest LGTM.

azat committed:
Rename fsync_tmp_directory to fsync_directories for Distributed engine

azat commented:
>Stress test (thread) — Server failed to start

The failure because of TSAN does not allow >64 locks per thread:
- https://github.com/google/sanitizers/issues/950#issuecomment-386711603
- https://github.com/llvm/llvm-project/blob/b02eab9058e58782fca32dd8b1e53c27ed93f866/compiler-rt/lib/sanitizer_common/sanitizer_deadlock_detector.h#L67

<details>

```
FATAL: ThreadSanitizer CHECK failed: /build/llvm-toolchain-11-11.0.1~++20201103062930+ef4ffcafbb2/compiler-rt/lib/sanitizer_common/sanitizer_deadlock_detector.h:67 "((n_all_locks_)) < (((sizeof(all_locks_with_contexts_)/sizeof((all_locks_with_contexts_)[0]))))" (0x40, 0x40)
==311==Launching Symbolizer process: /usr/bin/llvm-symbolizer-11 --inlining=true --default-arch=x86_64 
    #0 __tsan::TsanCheckFailed(char const*, int, char const*, unsigned long long, unsigned long long) <null> (clickhouse+0x8ceefa5)
    #1 __sanitizer::CheckFailed(char const*, int, char const*, unsigned long long, unsigned long long) <null> (clickhouse+0x8d05d4f)
    #2 __sanitizer::DeadlockDetectorTLS<__sanitizer::TwoLevelBitVector<1ul, __sanitizer::BasicBitVector<unsigned long> > >::addLock(unsigned long, unsigned long, unsigned int) <null> (clickhouse+0x8cfa4d6)
    #3 __sanitizer::DD::MutexAfterLock(__sanitizer::DDCallback*, __sanitizer::DDMutex*, bool, bool) <null> (clickhouse+0x8cf9a6e)
    #4 __tsan::MutexPostLock(__tsan::ThreadState*, unsigned long, unsigned long, unsigned int, int) <null> (clickhouse+0x8ced2f5)
    #5 pthread_mutex_lock <null> (clickhouse+0x8ca4878)
    #6 std::__1::__libcpp_mutex_lock(pthread_mutex_t*) obj-x86_64-linux-gnu/../contrib/libcxx/include/__threading_support:322:10 (clickhouse+0x18115549)
    #7 std::__1::mutex::lock() obj-x86_64-linux-gnu/../contrib/libcxx/src/mutex.cpp:33:14 (clickhouse+0x18115549)
    #8 std::__1::unique_lock<std::__1::mutex>::unique_lock(std::__1::mutex&) obj-x86_64-linux-gnu/../contrib/libcxx/include/__mutex_base:119:61 (clickhouse+0x8d71c99)
    #9 void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>) obj-x86_64-linux-gnu/../src/Common/ThreadPool.cpp:101:26 (clickhouse+0x8d71c99)
    #10 ThreadPoolImpl<std::__1::thread>::scheduleOrThrow(std::__1::function<void ()>, int, unsigned long) obj-x86_64-linux-gnu/../src/Common/ThreadPool.cpp:160:5 (clickhouse+0x8d72848)
    #11 ThreadFromGlobalPool::ThreadFromGlobalPool<void ThreadPoolImpl<ThreadFromGlobalPool>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda1'()>(void&&, void ThreadPoolImpl<ThreadFromGlobalPool>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda1'()&&...) obj-x86_64-linux-gnu/../src/Common/ThreadPool.h:162:38 (clickhouse+0x8d77ea7)
    #12 void ThreadPoolImpl<ThreadFromGlobalPool>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>) obj-x86_64-linux-gnu/../src/Common/ThreadPool.cpp:124:35 (clickhouse+0x8d74350)
    #13 ThreadPoolImpl<ThreadFromGlobalPool>::scheduleOrThrowOnError(std::__1::function<void ()>, int) obj-x86_64-linux-gnu/../src/Common/ThreadPool.cpp:148:5 (clickhouse+0x8d73e80)
    #14 DB::InterpreterSystemQuery::restartReplicas(DB::Context&) obj-x86_64-linux-gnu/../src/Interpreters/InterpreterSystemQuery.cpp:471:14 (clickhouse+0x12b75658)
    #15 DB::InterpreterSystemQuery::execute() obj-x86_64-linux-gnu/../src/Interpreters/InterpreterSystemQuery.cpp:364:13 (clickhouse+0x12b6fc84)
    #16 DB::executeQueryImpl(char const*, char const*, DB::Context&, bool, DB::QueryProcessingStage::Enum, bool, DB::ReadBuffer*) obj-x86_64-linux-gnu/../src/Interpreters/executeQuery.cpp:525:32 (clickhouse+0x12cff5de)
    #17 DB::executeQuery(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, DB::Context&, bool, DB::QueryProcessingStage::Enum, bool) obj-x86_64-linux-gnu/../src/Interpreters/executeQuery.cpp:868:30 (clickhouse+0x12cfde1e)
    #18 DB::TCPHandler::runImpl() obj-x86_64-linux-gnu/../src/Server/TCPHandler.cpp:260:24 (clickhouse+0x1352f2f2)
    #19 DB::TCPHandler::run() obj-x86_64-linux-gnu/../src/Server/TCPHandler.cpp:1417:9 (clickhouse+0x1353b7f7)
    #20 Poco::Net::TCPServerConnection::start() obj-x86_64-linux-gnu/../contrib/poco/Net/src/TCPServerConnection.cpp:43:3 (clickhouse+0x15ff82e2)
    #21 Poco::Net::TCPServerDispatcher::run() obj-x86_64-linux-gnu/../contrib/poco/Net/src/TCPServerDispatcher.cpp:112:19 (clickhouse+0x15ff8a0e)
    #22 Poco::PooledThread::run() obj-x86_64-linux-gnu/../contrib/poco/Foundation/src/ThreadPool.cpp:199:14 (clickhouse+0x161583d1)
    #23 Poco::(anonymous namespace)::RunnableHolder::run() obj-x86_64-linux-gnu/../contrib/poco/Foundation/src/Thread.cpp:55:11 (clickhouse+0x16156a9f)
    #24 Poco::ThreadImpl::runnableEntry(void*) obj-x86_64-linux-gnu/../contrib/poco/Foundation/src/Thread_POSIX.cpp:345:27 (clickhouse+0x16155407)
    #25 __tsan_thread_start_func <null> (clickhouse+0x8c8712c)
    #26 start_thread <null> (libpthread.so.0+0x9608)
    #27 clone <null> (libc.so.6+0x122292)
```

</details>

alexey-milovidov commented:
@azat We have not seen this issue for a long time.
What are these locks?

I doubt we can proceed with TSan failures.
Even if the failure is unrelated to this PR, it still has to be investigated.

azat commented:
>@azat We have not seen this issue for a long time.

@alexey-milovidov I guess that was just a luck, since due to parallel tests execution you can have >64 replicated tables at once

>What are these locks?

I believe that this is table lock, from InterpreterSystemQuery::tryRestartReplica

> I doubt we can proceed with TSan failures.
> Even if the failure is unrelated to this PR, it still has to be investigated.

#18897
